File: backend/app/schemas/athlete.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from app.models.enums import CoachAthleteLinkStatus, Gender

FOCUS_IMPORTANCE_MIN = 20
FOCUS_IMPORTANCE_MAX = 100


class AthleteOnboardingRequest(BaseModel):
    gender: Gender
    birth_date: date
    focus_strength: int = Field(ge=FOCUS_IMPORTANCE_MIN, le=FOCUS_IMPORTANCE_MAX)
    focus_flexibility: int = Field(ge=FOCUS_IMPORTANCE_MIN, le=FOCUS_IMPORTANCE_MAX)
    focus_endurance: int = Field(ge=FOCUS_IMPORTANCE_MIN, le=FOCUS_IMPORTANCE_MAX)
    focus_coordination: int = Field(ge=FOCUS_IMPORTANCE_MIN, le=FOCUS_IMPORTANCE_MAX)
    weight_target_min_kg: float | None = Field(default=None, ge=20, le=300)
    weight_target_max_kg: float | None = Field(default=None, ge=20, le=300)
    personal_goal_title: str | None = Field(default=None, max_length=200)
    personal_goal_target: float | None = Field(default=None, ge=0, le=1_000_000)

    @field_validator("birth_date")
    @classmethod
    def birth_date_in_past(cls, value: date) -> date:
        if value >= date.today():
            raise ValueError("Дата рождения должна быть в прошлом")
        return value

    @field_validator("personal_goal_title")
    @classmethod
    def normalize_goal_title(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None

    @model_validator(mode="after")
    def validate_ranges(self) -> "AthleteOnboardingRequest":
        if (
            self.weight_target_min_kg is not None
            and self.weight_target_max_kg is not None
            and self.weight_target_min_kg > self.weight_target_max_kg
        ):
            raise ValueError("Минимальный вес должен быть не больше максимального")

        has_title = self.personal_goal_title is not None
        has_target = self.personal_goal_target is not None
        if has_title and not has_target:
            raise ValueError("Укажите целевое значение личной цели")
        if has_target and not has_title:
            raise ValueError("Укажите название личной цели")

        return self


class AthleteProfileUpdateRequest(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=120)
    gender: Gender | None = None
    birth_date: date | None = None

    @field_validator("display_name")
    @classmethod
    def normalize_display_name(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            raise ValueError("Имя не может быть пустым")
        return stripped

    @field_validator("birth_date")
    @classmethod
    def birth_date_in_past(cls, value: date | None) -> date | None:
        if value is not None and value >= date.today():
            raise ValueError("Дата рождения должна быть в прошлом")
        return value


class JoinCoachRequest(BaseModel):
    invite_code: str = Field(min_length=4, max_length=32)


class AthleteCoachResponse(BaseModel):
    link_id: UUID
    coach_id: UUID
    display_name: str
    link_status: CoachAthleteLinkStatus
    sessions_balance: int = 0


class AthleteProfileResponse(BaseModel):
    display_name: str
    gender: Gender | None = None
    birth_date: date | None = None
    avatar_url: str | None = None
    timezone: str | None = None
    focus_strength: int | None = None
    focus_flexibility: int | None = None
    focus_endurance: int | None = None
    focus_coordination: int | None = None
    weight_target_min_kg: float | None = None
    weight_target_max_kg: float | None = None
    personal_goal_title: str | None = None
    personal_goal_target: float | None = None
    onboarding_completed_at: datetime | None = None

    model_config = {"from_attributes": True}

File: packages/shared/src/athlete.ts
export type Gender = "male" | "female";

export const GENDER_LABELS: Record<Gender, string> = {
  male: "Мужской",
  female: "Женский",
};

export const FOCUS_IMPORTANCE_MIN = 20;
export const FOCUS_IMPORTANCE_MAX = 100;
export const FOCUS_IMPORTANCE_DEFAULT = 50;

export type TrainingTrait = "strength" | "flexibility" | "endurance" | "coordination";

export const TRAINING_TRAIT_LABELS: Record<TrainingTrait, string> = {
  strength: "Сила",
  flexibility: "Гибкость",
  endurance: "Выносливость",
  coordination: "Координация",
};

export interface AthleteProfile {
  display_name: string;
  gender?: Gender | null;
  birth_date?: string | null;
  avatar_url?: string | null;
  timezone?: string | null;
  focus_strength?: number | null;
  focus_flexibility?: number | null;
  focus_endurance?: number | null;
  focus_coordination?: number | null;
  weight_target_min_kg?: number | null;
  weight_target_max_kg?: number | null;
  personal_goal_title?: string | null;
  personal_goal_target?: number | null;
  onboarding_completed_at?: string | null;
}

export interface AthleteOnboardingPayload {
  gender: Gender;
  birth_date: string;
  focus_strength: number;
  focus_flexibility: number;
  focus_endurance: number;
  focus_coordination: number;
  weight_target_min_kg?: number | null;
  weight_target_max_kg?: number | null;
  personal_goal_title?: string | null;
  personal_goal_target?: number | null;
}

export interface AthleteProfileUpdatePayload {
  display_name?: string;
  gender?: Gender;
  birth_date?: string;
}

export interface AthleteCoachLink {
  link_id: string;
  coach_id: string;
  display_name: string;
  avatar_url?: string | null;
  link_status: "pending" | "active" | "paused" | "ended";
  sessions_balance: number;
  sessions_completed: number;
}

export interface AthleteSessionsStats {
  sessions_completed: number;
}

export interface AthleteCompleteSessionResponse {
  link_id?: string | null;
  sessions_balance?: number | null;
  sessions_completed: number;
  activity_name?: string | null;
  duration_min?: number | null;
  effort?: number | null;
  effective_met?: number | null;
  load_met_minutes?: number | null;
  weight_kg_used?: number | null;
  calories_kcal?: number | null;
}

export interface AthleteCompleteSessionPayload {
  link_id?: string;
  without_coach?: boolean;
  activity_type_id: string;
  duration_min: number;
  effort: number;
}

/** Значение выбора «без тренера» в UI добавления тренировки. */
export const ATHLETE_WORKOUT_WITHOUT_COACH = "__without_coach__" as const;
export const ATHLETE_WORKOUT_WITHOUT_COACH_LABEL = "Без тренера";

export interface AthleteLastSession {
  entry_date: string;
  activity_name?: string | null;
  duration_min?: number | null;
  effort?: number | null;
  effective_met?: number | null;
  load_met_minutes?: number | null;
  weight_kg_used?: number | null;
  calories_kcal?: number | null;
  coach_display_name?: string | null;
}

export function formatAthleteLastSessionDate(entryDate: string): string {
  const date = new Date(`${entryDate}T00:00:00`);
  if (Number.isNaN(date.getTime())) return entryDate;
  return new Intl.DateTimeFormat("ru-RU", {
    day: "numeric",
    month: "long",
    year: "numeric",
  }).format(date);
}

export interface JoinCoachPayload {
  invite_code: string;
  claim_athlete_id?: string;
}

export type CoachAthleteSessionEntryKind = "credit" | "debit";

export interface CoachAthleteSessionHistoryEntry {
  id: string;
  kind: CoachAthleteSessionEntryKind;
  sessions_count: number;
  entry_date: string;
}

export const ATHLETE_APP_STATUS_LABELS = {
  withApp: "В приложении",
  withoutApp: "Без приложения",
} as const;

export function getAthleteAppStatusLabel(hasApp: boolean): string {
  return hasApp ? ATHLETE_APP_STATUS_LABELS.withApp : ATHLETE_APP_STATUS_LABELS.withoutApp;
}

export interface CoachAthleteSummary {
  athlete_id: string;
  display_name: string;
  has_app: boolean;
  avatar_url?: string | null;
  link_id: string;
  link_status: "pending" | "active" | "paused" | "ended";
  sessions_balance: number;
  sessions_completed: number;
  gender?: Gender | null;
  birth_date?: string | null;
  focus_strength?: number | null;
  focus_flexibility?: number | null;
  focus_endurance?: number | null;
  focus_coordination?: number | null;
  weight_target_min_kg?: number | null;
  weight_target_max_kg?: number | null;
  personal_goal_title?: string | null;
  personal_goal_target?: number | null;
  onboarding_completed_at?: string | null;
}

export interface CoachAthleteSessionsResponse {
  athlete_id: string;
  link_id: string;
  sessions_balance: number;
  sessions_completed: number;
}

export function isAthleteOnboardingComplete(profile: AthleteProfile | null | undefined): boolean {
  return Boolean(profile?.onboarding_completed_at);
}

export function clampFocusImportance(value: number): number {
  return Math.max(
    FOCUS_IMPORTANCE_MIN,
    Math.min(FOCUS_IMPORTANCE_MAX, Math.round(value)),
  );
}

export function calculateAge(birthDate: string | null | undefined, today = new Date()): number | null {
  if (!birthDate) return null;
  const born = new Date(`${birthDate}T00:00:00`);
  if (Number.isNaN(born.getTime())) return null;
  let age = today.getFullYear() - born.getFullYear();
  const monthDiff = today.getMonth() - born.getMonth();
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < born.getDate())) {
    age -= 1;
  }
  return age >= 0 ? age : null;
}

export function formatBirthDateDisplay(birthDate: string | null | undefined): string {
  if (!birthDate) return "—";
  const date = new Date(`${birthDate}T00:00:00`);
  if (Number.isNaN(date.getTime())) return birthDate;
  return new Intl.DateTimeFormat("ru-RU", {
    day: "numeric",
    month: "long",
    year: "numeric",
  }).format(date);
}
